File: Baekjun/9.4/1561. 놀이 공원/solution.swift
let NM = readLine()!.split(separator: " ").compactMap { Int($0) }
var N = NM[0], M = NM[1]
let A: [Int] = readLine()!.split(separator: " ").compactMap { Int($0) }

if M >= N {
    print(N)
} else {
    var l = 1, r = A.max()! * N

    var t = r
    while r > l + 1 {
        let time: Int = (r + l) / 2

        if A.reduce(M, { result, nextValue in
            result + (time / nextValue)
        }) >= N {
            r = time
            t = min(t, r)
        } else {
            l = time
        }
    }

    var cnt: Int = A.reduce(M) { acc, cur in
        acc + (t - 1) / cur
    }

    for i in 0 ..< M {
        if t % A[i] == 0 { cnt += 1 }

        if cnt == N {
            print(i + 1)
            break
        }
    }
}

File: Baekjun/9.4/1561. 놀이 공원/solution.py
from functools import reduce

# N: 사람 수, M: 놀이기구 수
N, M = map(int, input().split())
A = list(map(int, input().split()))

if N <= M:
    print(N)
    exit(1)

l, r = 1, max(A) * N

tmp = r
while r > l + 1:
    time = (l + r) // 2
    if reduce(lambda acc, cur: acc + time // cur, A, M) >= N:
        r = time
        tmp = min(tmp, r)
    else:
        l = time

cnt = reduce(lambda acc, cur: acc + (tmp - 1) // cur, A, M)

for i in range(M):
    if not tmp % A[i]:
        cnt += 1
    if cnt == N:
        print(i + 1)
        break
